game-app/game-core/src/main/java/games/strategy/triplea/ai/pro/ProCombatMoveAi.java

```diff
@@ -1197,6 +1156,91 @@ && canAirSafelyLandAfterAttack(unit, t)) {
+        .entrySet()
+        .forEach(
+            bomberEntry -> {
+              final Unit bomber = bomberEntry.getKey();
+              if (alreadyAttackedWithUnits.contains(bomber)) {
+                return; // already attacked bombers cannot move
+              }
+              Set<Territory> bomberTargetTerritories = bomberEntry.getValue();
+              if (raidsMayBePrecededByAirBattles) {
+                bomberTargetTerritories =
+                    new HashSet(
+                        CollectionUtils.getMatches(
+                            bomberTargetTerritories,
+                            terr ->
+                                AirBattle.territoryCouldPossiblyHaveAirBattleDefenders(
+                                    terr, player, data, true)));
+              }
+              determineBestBombingAttackForBomber(
+                  attackMap, sortedUnitAttackOptions, bomberTargetTerritories, bomber);
+            });
+  }
+
+  /**
+   * Determines the best bombing attack territory for a single {@code bomber}.
+   *
+   * @param attackMap Attack map
+   * @param sortedUnitAttackOptions Sorted attack options
+   * @param bomberTargetTerritories Target territories
+   * @param bomber Bomber unit
+   */
+  private void determineBestBombingAttackForBomber(
+      final Map<Territory, ProTerritory> attackMap,
+      final Map<Unit, Set<Territory>> sortedUnitAttackOptions,
+      final Set<Territory> bomberTargetTerritories,
+      final Unit bomber) {
+    final Predicate<Unit> bombingTargetMatch =
+        Matches.unitCanProduceUnitsAndCanBeDamaged()
+            .and(Matches.unitIsLegalBombingTargetBy(bomber));
+    Optional<Territory> maxBombingTerritory = Optional.empty();
+    int maxBombingScore = MIN_BOMBING_SCORE;
+    for (final Territory t : bomberTargetTerritories) {
+      final List<Unit> targetUnits = t.getUnitCollection().getMatches(bombingTargetMatch);
+      if (!targetUnits.isEmpty() && canAirSafelyLandAfterAttack(bomber, t)) {
+        final int noAaBombingDefense =
+            t.anyUnitsMatch(Matches.unitIsAaForBombingThisUnitOnly()) ? 0 : 1;
+        int maxDamageProduction = TerritoryAttachment.getProduction(t);
+        int neededDamageUnits = 0;
+        final List<Unit> existingAttackingBombers = attackMap.get(t).getBombers();
+        final List<Unit> sameTargetBombers = new ArrayList<>();
+        for (Unit target : targetUnits) {
+          neededDamageUnits += target.getHowMuchMoreDamageCanThisUnitTake(t);
+          for (Unit existingBomber : existingAttackingBombers) {
+            if (!sameTargetBombers.contains(existingBomber)
+                && Matches.unitIsLegalBombingTargetBy(existingBomber).test(target)) {
+              sameTargetBombers.add(existingBomber);
+            }
+          }
+        }
+        // assume each other bomber causes a damage of 3
+        final int remainingDamagePotential =
+            maxDamageProduction - neededDamageUnits - 3 * sameTargetBombers.size();
```

Comment:
I don't understand the `maxDamageProduction  - neededDamageUnits` expression.

That sounds to me that the territory production value gets reduced by how much damage units can still take.
That's not right, is it?

I would expect it should either be the sum of the two (e.g. max bombing damage = production + damage that individual factories can take), or choosing one or the other, depending on game rules, or the max or min of the two. 
Subtracting just seems not right to me.

Can we verify with the engine code where damage is actually determined from bombing?

Reply:
It should be `maxDamageProduction + neededDamageUnits` and I corrected it.

Concerning the damage logic from bombing you can check StrategicBombingRaidBattle especially variable `bombingRaidDamage` and call of method `getTarget`.

Reply:
Thanks. The relevant code seems to be:

```
      int damageLimit = TerritoryAttachment.getProduction(battleSite);
      final boolean limitDamage =
          Properties.getWW2V2(gameData.getProperties())
              || Properties.getLimitRocketAndSbrDamageToProduction(gameData.getProperties());

        if (limitDamage) {
          costThisUnit = Math.min(costThisUnit, damageLimit);
        }
```

and:

```java
      // Limit PUs lost if we would like to cap PUs lost at territory value
      if (Properties.getPuCap(gameData.getProperties())
          || Properties.getLimitSbrDamagePerTurn(gameData.getProperties())) {
        final int alreadyLost = gameData.getMoveDelegate().pusAlreadyLost(battleSite);
        final int limit = Math.max(0, damageLimit - alreadyLost);
        cost = Math.min(cost, limit);
        if (!targets.isEmpty()) {
          for (final Unit u : bombingRaidDamage.keySet()) {
            if (bombingRaidDamage.getInt(u) > limit) {
              bombingRaidDamage.put(u, limit);
            }
          }
        }
      }
```

So if those properties are in place, then we should consider that `damageLimit` (which is `TerritoryAttachment.getProduction(battleSite)`) should be the max, both for the overall cost and to the cost to each target unit.

Additionally, right after there's code that also enforces the `damageLimit` per unit if `Properties.getDamageFromBombingDoneToUnitsInsteadOfTerritories()` is set.

Importantly, if `Properties.getDamageFromBombingDoneToUnitsInsteadOfTerritories()` is false, `bombingRaidDamage` isn't used.

That means, we should take that account in the AI code you're changing. Specifically, `neededDamageUnits` should be 0 in that case.

Otherwise, the other properties should be used to cap the damage per bomber, so I think instead of `3`, it should be `Math.min(3,  maxDamageProduction)`, if `Properties.getPuCap(gameData.getProperties()) || Properties.getLimitSbrDamagePerTurn(gameData.getProperties()))`.
Note: The 3 is still bogus and would ideally be calculated from the expected value of the bomber's dice, but that can be left as a TODO (3 is close to the expected value of 3.5 for a regular d6). Adding the `Math.min()` seems a simple enough change that I think it's worthwhile to do here.

Reply:
One other note:

Really what we care about is the application of the limit - and it is pretty peculiar.

if `Properties.getDamageFromBombingDoneToUnitsInsteadOfTerritories()` is set, then the limit is per unit, so it's straight up  `maxDamageProduction * targetUnits.size()`, *if* the limit should be applied (based on properties).

If not, then the target units are not relevant, and it's just the overall limit (either infinity or the production value, depending on properties).

Reply:
I agree to the case where the property is _not_ set (then only the territory is relevant, meaning its production value/infinitive).
But in the case where the property is set, each unit needs to be checked how much damage it can take and this is added to the territory part (again its production value/infinitive). This is not the same as your formula `maxDamageProduction * targetUnits.size()`.
If that is true the only thing I see that could be changed is the target units part if the property is _not_ set.
Did you wanted to express a different part or did I get you right?

Reply:
Yes, sorry you're right. If the property is set, we actually do `damageLimit = current.getHowMuchMoreDamageCanThisUnitTake(battleSite)` - so it's not based on `maxDamageProduction`. My mistake.

Reply:
Then this part is resolved, right?